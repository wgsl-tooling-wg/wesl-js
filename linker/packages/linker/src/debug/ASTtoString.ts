import {
  AbstractElem,
  ExpressionElem,
  ModuleElem,
  TypeRefElem,
  TypeTemplateParameter,
} from "../AbstractElems.ts";
import { importToString } from "./ImportToString.ts";
import { LineWrapper } from "./LineWrapper.ts";

const maxLineLength = 150;

export function astToString(elem: AbstractElem, indent = 0): string {
  const { kind, contents } = elem as ModuleElem;
  const str = new LineWrapper(indent, maxLineLength);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (contents) {
    childStrings = contents.map(e => astToString(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

// TODO DRY with above
export function elemToString(elem: AbstractElem): string {
  const { kind } = elem as ModuleElem;
  const str = new LineWrapper(0, maxLineLength);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }
  return str.result;
}

function addElemFields(elem: AbstractElem, str: LineWrapper): void {
  addTextFields(elem, str) ||
    addVarishFields(elem, str) ||
    addStructFields(elem, str) ||
    addStructMemberFields(elem, str) ||
    addNameFields(elem, str) ||
    addMemberRef(elem, str) ||
    addFnFields(elem, str) ||
    addAliasFields(elem, str) ||
    addAttributeFields(elem, str) ||
    addExpressionFields(elem, str) ||
    addTypeRefFields(elem, str) ||
    addSynthetic(elem, str) ||
    addImport(elem, str) ||
    addRefIdent(elem, str) ||
    addTypedDeclIdent(elem, str) ||
    addDeclIdent(elem, str);
}

function addVarishFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  const { kind } = elem;
  if (
    kind === "var" ||
    kind === "let" ||
    kind === "gvar" ||
    kind === "const" ||
    kind === "override"
  ) {
    addTypedDeclIdent(elem.name, str);
    return true;
  }
}

function addTextFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "text") {
    const { srcModule, start, end } = elem;
    str.add(` '${srcModule.src.slice(start, end)}'`);
    return true;
  }
}

function addRefIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "ref") {
    str.add(" " + elem.ident.originalName);
    return true;
  }
}

function addMemberRef(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "memberRef") {
    str.add(` ${elem.name.ident.originalName}.${elem.member.name}`);
    return true;
  }
}

function addDeclIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "decl") {
    const { ident } = elem;
    str.add(" %" + ident.originalName);
    return true;
  }
}

function addTypedDeclIdent(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "typeDecl") {
    const { decl, typeRef } = elem;
    str.add(" %" + decl.ident.originalName);
    if (typeRef) {
      str.add(" : " + typeRefElemToString(typeRef));
    }
    return true;
  }
}

function addAliasFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRefElemToString(typeRef));
    return true;
  }
}

function addStructFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
    return true;
  }
}

function addStructMemberFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "member") {
    const { name, typeRef, attributes } = elem;
    if (attributes) {
      str.add(" " + attributes.map(a => "@" + (a && a.name)).join(" "));
    }
    str.add(" " + name.name);
    str.add(": " + typeRefElemToString(typeRef));
    return true;
  }
}

function addImport(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "import") {
    str.add(" " + importToString(elem.imports));
    return true;
  }
}

function addSynthetic(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "synthetic") {
    str.add(` '${elem.text}'`);
    return true;
  }
}

function addAttributeFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "attribute") {
    const { name, params } = elem;
    str.add(" @" + name);
    if (params) {
      str.add("(");
      str.add(params.map(expressionToString).join(", "));
      str.add(")");
    }
    return true;
  }
}

function addNameFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "name") {
    str.add(" " + elem.name);
    return true;
  }
}

function addFnFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "fn") {
    const { name, params, returnType, fnAttributes } = elem;

    str.add(" " + name.ident.originalName);

    str.add("(");
    const paramStrs = params
      .map(
        (
          p, // TODO DRY
        ) => {
          const { name } = p;
          const { originalName } = name.decl.ident;
          const typeRef = typeRefElemToString(name.typeRef!);
          return originalName + ": " + typeRef;
        },
      )
      .join(", ");
    str.add(paramStrs);
    str.add(")");

    fnAttributes?.forEach(a => {
      str.add(" @" + a?.name);
      if (a?.params) {
        str.add("(");
        str.add(a.params.map(expressionToString).join(", "));
        str.add(")");
      }
    });

    if (returnType) {
      str.add(" -> " + typeRefElemToString(returnType));
    }

    return true;
  }
}

function addTypeRefFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "type") {
    const { name } = elem;
    const nameStr = typeof name === "string" ? name : name.originalName;
    str.add(" " + nameStr);

    if (elem.templateParams !== undefined) {
      const paramStrs = elem.templateParams
        .map(templateParamToString)
        .join(", ");
      str.add("<" + paramStrs + ">");
    }
    return true;
  }
}
function addExpressionFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "expression") {
    const contents = elem.contents
      .map(e => {
        if (e.kind === "text") {
          return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
        } else {
          return elemToString(e);
        }
      })
      .join(" ");
    str.add(" " + contents);
    return true;
  }
}

function expressionToString(elem: ExpressionElem): string {
  const contents = elem.contents
    .map(e => {
      if (e.kind === "text") {
        return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
      } else {
        return elemToString(e);
      }
    })
    .join(" ");
  return contents;
}

function templateParamToString(p: TypeTemplateParameter): string {
  if (typeof p === "string") {
    return p;
  } else if (p.kind === "type") {
    return typeRefElemToString(p);
  } else if (p.kind === "expression") {
    return expressionToString(p);
  } else {
    console.log("unknown template parameter type", p);
    return "??";
  }
}

function typeRefElemToString(elem: TypeRefElem): string {
  if (!elem) return "?type?";
  const { name } = elem;
  const nameStr = typeof name === "string" ? name : name.originalName;

  let params = "";
  if (elem.templateParams !== undefined) {
    const paramStrs = elem.templateParams.map(templateParamToString).join(", ");
    params = "<" + paramStrs + ">";
  }
  return nameStr + params;
}
