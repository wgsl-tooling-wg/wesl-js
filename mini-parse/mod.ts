export * from "./MatchingLexer.ts";
export * from "./Parser.ts";
export * from "./ParserCombinator.ts";
export * from "./ParserLogging.ts";
export * from "./ParserTracing.ts";
export * from "./SrcMap.ts";
export * from "./TokenMatcher.ts";
