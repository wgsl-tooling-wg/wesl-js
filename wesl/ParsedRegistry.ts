import { parseSrcModule, type WeslAST } from "./ParseWESL.ts";
import { normalize, noSuffix } from "./PathUtil.ts";
import { resetScopeIds, type SrcModule } from "./Scope.ts";
import type { WeslBundle } from "./WeslBundle.ts";

export interface ParsedRegistry {
  modules: Record<string, WeslAST>; // key is module path, e.g. "rand_pkg::foo::bar"
}

export function parsedRegistry(): ParsedRegistry {
  resetScopeIds(); // for debug
  return { modules: {} };
}

/** for debug */
export function registryToString(registry: ParsedRegistry): string {
  return `modules: ${[...Object.keys(registry.modules)]}`;
}

/** Look up a module with a flexible selector.
 *    :: separated module path,   package::util
 *    / separated file path       ./util.wesl (or ./util)
 *          - note: a file path should not include a weslRoot prefix, e.g. not ./shaders/util.wesl
 *    simpleName                  util
 */
export function selectModule(
  parsed: ParsedRegistry,
  selectPath: string,
  packageName = "package",
): WeslAST | undefined {
  // dlog({reg: [...Object.keys(parsed.modules)]});
  let modulePath: string;
  if (selectPath.includes("::")) {
    modulePath = selectPath;
  } else if (
    selectPath.includes("/") ||
    selectPath.endsWith(".wesl") ||
    selectPath.endsWith(".wgsl")
  ) {
    modulePath = fileToModulePath(selectPath, packageName);
  } else {
    modulePath = packageName + "::" + selectPath;
  }

  return parsed.modules[modulePath];
}

/**
 * @param srcFiles    map of source strings by file path
 *                    key is '/' separated relative path (relative to srcRoot, not absolute file path )
 *                    value is wesl source string
 * @param registry    add parsed modules to this registry
 * @param packageName name of package
 */
export function parseIntoRegistry(
  srcFiles: Record<string, string>,
  registry: ParsedRegistry,
  packageName: string = "package",
  debugWeslRoot?: string,
): void {
  if (debugWeslRoot === undefined) {
    debugWeslRoot = "";
  } else if (!debugWeslRoot.endsWith("/")) {
    debugWeslRoot += "/";
  }
  const srcModules: SrcModule[] = Object.entries(srcFiles).map(
    ([filePath, src]) => {
      const modulePath = fileToModulePath(filePath, packageName);
      return { modulePath, debugFilePath: debugWeslRoot + filePath, src };
    },
  );
  srcModules.forEach((mod) => {
    const parsed = parseSrcModule(mod, undefined);
    if (registry.modules[mod.modulePath]) {
      throw new Error(`duplicate module path: '${mod.modulePath}'`);
    }
    registry.modules[mod.modulePath] = parsed;
  });
}

export function parseLibsIntoRegistry(
  libs: WeslBundle[],
  registry: ParsedRegistry,
): void {
  libs.forEach(({ modules, name }) =>
    parseIntoRegistry(modules, registry, name)
  );
}

const libRegex = /^lib\.w[eg]sl$/i;

/** convert a file path (./foo/bar.wesl)
 *  to a module path (package::foo::bar) */
function fileToModulePath(filePath: string, packageName: string): string {
  if (filePath.includes("::")) {
    // already a module path
    return filePath;
  }
  if (packageName !== "package" && libRegex.test(filePath)) {
    // special case for lib.wesl files in external packages
    return packageName;
  }

  const strippedPath = noSuffix(normalize(filePath));
  const moduleSuffix = strippedPath.replaceAll("/", "::");
  const modulePath = packageName + "::" + moduleSuffix;
  return modulePath;
}
