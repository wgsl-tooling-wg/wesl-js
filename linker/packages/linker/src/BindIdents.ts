import { debugNames } from "mini-parse";
import { DeclarationElem } from "./AbstractElems2.ts";
import { FlatImport } from "./FlattenTreeImport.ts";
import { ParsedRegistry2 } from "./ParsedRegistry2.ts";
import { DeclIdent, exportDecl, RefIdent, Scope } from "./Scope.ts";
import { identToString, scopeIdentTree } from "./ScopeLogging.ts";
import { stdFn, stdType } from "./TraverseRefs.ts";
import { last, overlapTail } from "./Util.ts";
import { dlog } from "berry-pretty";
import { flatImports, WeslAST } from "./ParseWESL.ts";

/**
 * Bind active reference idents to declaration Idents by mutating the refersTo: field
 * Also in this pass, set the mangledName: field for all active global declaration idents.
 *
 * @param parsed
 * @param conditions  only bind to/from idents that are valid with the current condition set
 * @return any new declaration elements found (they will need to be emitted)
 */
export function bindIdents(
  ast: WeslAST,
  parsed: ParsedRegistry2,
  conditions: Record<string, any>,
): DeclarationElem[] {
  /* 
    For each module's scope, search through the scope tree to find all ref idents
      - For each ref ident, search up the scope tree to find a matching decl ident
      - If no local match is found, check for partial matches with import statements
        - combine ident with import statement to match a decl in exporting module

    As global decl idents are found, mutate their mangled name to be globally unique.
*/
  const { rootScope } = ast;

  const globalNames = new Set<string>();
  const knownDecls = new Set<DeclIdent>();
  rootScope.idents.forEach(ident => {
    if (ident.kind === "decl") {
      ident.mangledName = ident.originalName;
      globalNames.add(ident.originalName);
      knownDecls.add(ident);
    }
  });

  const bindContext = {
    parsed,
    conditions,
    knownDecls,
    foundScopes: new Set<Scope>(),
    globalNames,
  };
  const decls = bindIdentsRecursive(rootScope, bindContext);
  return decls.flatMap(d =>
    d.declElem && isGlobal(d.declElem) ? [d.declElem] : [],
  );
}

interface BindContext {
  parsed: ParsedRegistry2;
  conditions: Record<string, any>;
  knownDecls: Set<DeclIdent>; // decl idents discovered so far
  foundScopes: Set<Scope>; // save work by not processing scopes multiple times
  globalNames: Set<string>; // root level names  used so far
}

/**
 * Recursively bind references to declarations in this scope and
 * any child scopes referenced by these declarations.
 * Uses a hash set of found declarations to avoid duplication
 * @ return any new declarations found
 */
function bindIdentsRecursive(
  scope: Scope,
  bindContext: BindContext,
): DeclIdent[] {
  // early exist if we've processed this scope before
  const { foundScopes } = bindContext;
  if (foundScopes.has(scope)) return [];
  foundScopes.add(scope);

  // dlog(scopeIdentTree(scope));

  const { parsed, conditions } = bindContext;
  const { globalNames, knownDecls } = bindContext;
  const newDecls: DeclIdent[] = []; // new decl idents to process (and return)

  scope.idents.forEach((ident, i) => {
    // dlog(`--- considering ident ${identToString(ident)}`);
    if (ident.kind === "ref") {
      if (!ident.refersTo && !ident.std) {
        if (stdWgsl(ident.originalName)) {
          ident.std = true;
        } else {
          let foundDecl =
            findDeclInModule(scope, ident, i) ?? findDeclImport(ident, parsed);

          if (foundDecl && !knownDecls.has(foundDecl)) {
            // dlog(
            //   `  > found decl: ${identToString(foundDecl)} known: ${knownDecls.has(foundDecl)}`,
            // );
            setDisplayName(ident.originalName, foundDecl, globalNames);
            // dlog(`  > queuing new decl: ${identToString(foundDecl)}`);
            if (foundDecl.declElem && isGlobal(foundDecl.declElem)) {
              newDecls.push(foundDecl);
            }
            knownDecls.add(foundDecl);
          }
          bindRefToDecl(ident, foundDecl, knownDecls);
        }
      }
    }
  });

  // follow references from child scopes
  const newFromChildren = scope.children.flatMap(child => {
    // dlog("newFromChildren", { childScope: scopeIdentTree(child) });
    return bindIdentsRecursive(child, bindContext);
  });
  // console.log(
  //   "new from children",
  //   newFromChildren.map(d => identToString(d)),
  // );

  // follow references from referenced declarations
  const newFromRefs = newDecls.flatMap(decl => {
    // dlog("newFromRefs", {
    //   decl: identToString(decl),
    //   declScope: scopeIdentTree(decl.scope),
    // });
    if (debugNames && !decl.scope) {
      console.log(`--- decl ${identToString(decl)} has no scope`);
      return [];
    }
    return bindIdentsRecursive(decl.scope, bindContext);
  });
  // console.log(
  //   "new from refs",
  //   newFromRefs.map(d => identToString(d)),
  // );

  return [newDecls, newFromChildren, newFromRefs].flat();
}

function setDisplayName(
  proposedName: string,
  decl: DeclIdent,
  globalNames: Set<string>,
): void {
  if (!decl.mangledName) {
    // if (!decl.declElem) {
    //   console.log(
    //     `--- decl ident ${identToString(decl)} has no declElem attached`,)
    // } else
    if (decl.declElem && isGlobal(decl.declElem)) {
      decl.mangledName = declUniqueName(proposedName, globalNames);
      // dlog(`  > mangle global decl: ${identToString(decl)}`);
    } else {
      // dlog(`  > no-mangle local decl: ${identToString(decl)}`);
      decl.mangledName = decl.originalName;
    }
  }
}

function bindRefToDecl(
  ident: RefIdent,
  foundDecl: DeclIdent | undefined,
  knownDecls: Set<DeclIdent>,
) {
  // dlog({ ident: identToString(ident), foundDecl: identToString(foundDecl) });
  if (foundDecl) {
    ident.refersTo = foundDecl;

    knownDecls.add(foundDecl);
  } else {
    // TODO log error with source position
    console.log(`--- unresolved ident: ${ident.originalName}`);
  }
}

function stdWgsl(name: string): boolean {
  return stdType(name) || stdFn(name);
}

/** search earlier in the scope and in parent scopes to find a matching decl ident */
function findDeclInModule(
  scope: Scope,
  ident: RefIdent,
  identDex: number = scope.idents.length,
): DeclIdent | undefined {
  const { idents, parent } = scope;
  const { originalName } = ident;

  // dlog(`  |> findDeclInModule ${originalName}`);
  // see if the declaration is in this scope
  for (let i = identDex - 1; i >= 0; i--) {
    const checkIdent = idents[i];
    const { kind } = checkIdent;
    if (kind === "decl" && originalName === checkIdent.originalName) {
      // dlog(`  |> found decl in scope: ${identToString(checkIdent)}`);
      return checkIdent;
    }
  }

  // recurse to check all idents in parent scope
  if (parent) {
    return findDeclInModule(parent, ident);
  }
}

/** Match a reference identifier to a declaration in
 * another module via an import statement */
function findDeclImport(
  refIdent: RefIdent,
  parsed: ParsedRegistry2,
): DeclIdent | undefined {
  // dlog(identToString(refIdent), { ast: !!refIdent.ast });
  const flatImps = flatImports(refIdent.ast);

  // find module path by combining identifer reference with import statement
  const modulePathParts = matchingImport(refIdent, flatImps); // module path in array form

  if (modulePathParts) {
    return findExport(modulePathParts, parsed);
  }
}

/** using the flattened import array, find an import that matches a provided identifier */
function matchingImport(
  ident: RefIdent,
  flatImports: FlatImport[],
): string[] | undefined {
  const identParts = ident.originalName.split("::");
  for (const flat of flatImports) {
    const impTail = overlapTail(flat.importPath, identParts);
    if (impTail) {
      return [...flat.modulePath, ...impTail];
    }
  }
}

/** @return an exported root element for the provided path */
function findExport(
  modulePathParts: string[],
  parsed: ParsedRegistry2,
): DeclIdent | undefined {
  const legacyConvert = modulePathParts.map(p => (p === "." ? "package" : p)); // TODO rm after we update grammar
  const modulePath = legacyConvert.slice(0, -1).join("::");
  const module = parsed.modules[modulePath];
  if (!module) {
    console.log(
      `ident ${modulePathParts.join("::")} in import statement, but module not found`,
    );
  }

  return exportDecl(module.rootScope, last(modulePathParts)!);
}

/** return mangled name for decl ident,
 *  mutating the Ident to remember mangled name if it hasn't yet been determined */
export function declUniqueName(
  proposedName: string,
  rootNames: Set<string>,
): string {
  const displayName = uniquifyName(proposedName, rootNames);
  rootNames.add(displayName);

  return displayName;
}

/** construct global unique name for use in the output */
function uniquifyName(proposedName: string, rootNames: Set<string>): string {
  let renamed = proposedName;
  let conflicts = 0;

  // create a unique name
  while (rootNames.has(renamed)) {
    renamed = proposedName + conflicts++;
  }

  // dlog({ proposedName, renamed });
  return renamed;
}

export function isGlobal(elem: DeclarationElem): boolean {
  return ["alias", "const", "override", "fn", "struct", "gvar"].includes(
    elem.kind,
  );
}
