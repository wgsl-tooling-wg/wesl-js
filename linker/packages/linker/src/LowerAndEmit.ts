import { dlog } from "berry-pretty";
import { SrcMap, SrcMapBuilder } from "mini-parse";
import { AbstractElem2, IdentElem, TextElem } from "./AbstractElems2.ts";
import { Conditions, DeclIdent, Ident, RefIdent } from "./Scope.ts";

/** passed to the emitters */
interface EmitContext {
  srcMap: SrcMapBuilder; // constructing the linked output
  conditions: Conditions; // settings for conditional compilation
}

/** traverse the AST, starting from root elements, emitting wgsl for each */
export function lowerAndEmit(
  rootElems: AbstractElem2[],
  conditions: Conditions,
): SrcMap {
  const srcMap = new SrcMapBuilder();
  const emitContext: EmitContext = { conditions, srcMap };
  lowerAndEmitRecursive(rootElems, emitContext);
  return srcMap.build();
}

function lowerAndEmitRecursive(
  elems: AbstractElem2[],
  emitContext: EmitContext,
): void {
  const validElems = elems.filter(e =>
    conditionsValid(e, emitContext.conditions),
  );
  validElems.forEach(e => lowerAndEmitElem(e, emitContext));
}

export function lowerAndEmitElem(e: AbstractElem2, ctx: EmitContext): void {
  switch (e.kind) {
    case "text":
      return emitText(e, ctx);
    case "ident":
      return emitIdent(e, ctx);
    case "var":
    case "module":
    case "alias":
    case "override":
    case "const":
      return emitContents(e, ctx);
    default:
      dlog("ugh");
      throw new Error(`NYI emit elem kind: ${e.kind}`);
  }
}

export function emitText(e: TextElem, ctx: EmitContext): void {
  ctx.srcMap.addCopy(e.src, e.start, e.end);
}

export function emitContents(
  elem: AbstractElem2 & { contents: AbstractElem2[] },
  ctx: EmitContext,
): void {
  elem.contents.forEach(e => lowerAndEmitElem(e, ctx));
}

export function emitIdent(e: IdentElem, ctx: EmitContext): void {
  if ((e.ident as RefIdent).std) {
    ctx.srcMap.add(e.ident.originalName, e.src, e.start, e.end);
  } else {
    const declIdent = findDecl(e.ident);
    const mangledName = declIdent.mangledName!; // mangled name was set in binding step
    ctx.srcMap.add(mangledName, e.src, e.start, e.end);
  }
}

/** trace through refersTo links in reference Idents until we find the declaration
 * expects that bindIdents has filled in all refersTo: links
 */
export function findDecl(ident: Ident): DeclIdent {
  let i: Ident | undefined = ident;
  do {
    if (i.kind === "decl") {
      return i;
    }
    i = i.refersTo;
  } while (i);

  throw new Error(`unresolved ident: ${ident.originalName}`);
}

/** check if the element is visible with the current current conditional compilation settings */
export function conditionsValid(
  elem: AbstractElem2,
  conditions: Conditions,
): boolean {
  return true;
}
