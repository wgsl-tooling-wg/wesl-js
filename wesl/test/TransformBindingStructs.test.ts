import { SrcMapBuilder } from "@wesl/mini-parse";
import { expectTrimmedMatch } from "@wesl/mini-parse/vitest-util";
import { expect } from "@std/expect";
import { bindIdents } from "../BindIdents.ts";
import { astToString } from "../debug/ASTtoString.ts";
import { lowerAndEmit } from "../LowerAndEmit.ts";
import { parsedRegistry } from "../ParsedRegistry.ts";
import {
  bindingStructsPlugin,
  findRefsToBindingStructs,
  lowerBindingStructs,
  markBindingStructs,
  transformBindingReference,
  transformBindingStruct,
} from "../TransformBindingStructs.ts";
import { linkTestOpts, parseTest } from "./TestUtil.ts";
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("markBindingStructs true", () => {
  const src = `
    struct Bindings {
      @group(0) @binding(0) particles: ptr<storage, array<f32>, read_write>, 
    }
  `;

  const ast = parseTest(src);
  const structs = markBindingStructs(ast.moduleElem);
  expect(structs.length).toBe(1);
  expect(structs[0].bindingStruct).toBe(true);
});

Deno.test("markBindingStructs false", () => {
  const src = `
    struct Bindings {
      particles: ptr<storage, array<f32>, read_write>, 
    }
  `;

  const ast = parseTest(src);
  const structs = markBindingStructs(ast.moduleElem);
  expect(structs.length).toBe(0);
});

Deno.test("transformBindingStruct", async (t) => {
  const src = `
    struct Bindings {
      @group(0) @binding(0) particles: ptr<storage, array<f32>, read_write>, 
    }
  `;

  const rootAst = parseTest(src);
  bindIdents({ rootAst, registry: parsedRegistry() });
  const bindingStruct = markBindingStructs(rootAst.moduleElem)[0];
  const newVars = transformBindingStruct(bindingStruct, new Set());

  const srcBuilder = new SrcMapBuilder({ text: rootAst.srcModule.src });
  lowerAndEmit(srcBuilder, newVars, {});
  const linked = SrcMapBuilder.build([srcBuilder]).dest.text;
  await assertSnapshot(t, linked);
});

Deno.test("findRefsToBindingStructs", async (t) => {
  const src = `
    struct Bindings {
      @group(0) @binding(0) particles: ptr<storage, array<f32>, read_write>, 
    }
    struct NotBindings { a: i32 }
    var y: NotBindings;

    fn main(b: Bindings) {
      let x = b.particles;
      let z = y.a;
    }
  `;

  const rootAst = parseTest(src);
  bindIdents({ rootAst, registry: parsedRegistry() });
  markBindingStructs(rootAst.moduleElem)[0];
  const found = findRefsToBindingStructs(rootAst.moduleElem);
  expect(found.length).toBe(1);
  const foundAst = astToString(found[0].memberRef);
  await assertSnapshot(t, foundAst);
});

Deno.test("transformBindingReference", async (t) => {
  const src = `
    struct Bindings {
      @group(0) @binding(0) particles: ptr<storage, array<f32>, read_write>, 
    }
    fn main(b: Bindings) {
      let x = b.particles;
    }
  `;

  const rootAst = parseTest(src);
  bindIdents({ rootAst, registry: parsedRegistry() });
  const bindingStruct = markBindingStructs(rootAst.moduleElem)[0];
  transformBindingStruct(bindingStruct, new Set());
  const found = findRefsToBindingStructs(rootAst.moduleElem);
  expect(found.length).toBe(1);
  const { memberRef, struct } = found[0];
  const synthElem = transformBindingReference(memberRef, struct);
  const synthAst = astToString(synthElem);
  await assertSnapshot(t, synthAst);
});

Deno.test("lower binding structs", async (t) => {
  const src = `
    struct Bindings {
      @group(0) @binding(0) particles: ptr<storage, array<f32>, read_write>, 
    }
    fn main(b: Bindings) {
      let x = b.particles;
    }
  `;

  const expected = `
@group(0) @binding(0) var<storage, read_write> particles : array<f32>;
       
    fn main() {
      let x = particles;
    }
  `;
  const rootAst = parseTest(src);
  const { globalNames } = bindIdents({ rootAst, registry: parsedRegistry() });
  const tAst = { ...rootAst, globalNames, notableElems: {} };
  const lowered = lowerBindingStructs(tAst);

  const loweredAst = astToString(lowered.moduleElem);
  await assertSnapshot(t, loweredAst);

  const srcBuilder = new SrcMapBuilder({ text: lowered.srcModule.src });
  lowerAndEmit(srcBuilder, [lowered.moduleElem], {}, false);
  const linked = SrcMapBuilder.build([srcBuilder]).dest.text;
  expectTrimmedMatch(linked, expected);
});

Deno.test("lower binding structs with conflicting root name", async () => {
  const src = `
    struct Bindings {
      @group(0) @binding(0) particles: ptr<storage, array<f32>, read_write>, 
    }
    const particles = 7;
    fn main(b: Bindings) {
      let x = b.particles;
    }
  `;

  const expected = `
@group(0) @binding(0) var<storage, read_write> particles0 : array<f32>;
       
    const particles = 7;
    fn main() {
      let x = particles0;
    }
  `;

  const opts = { config: { plugins: [bindingStructsPlugin()] } };
  const linked = await linkTestOpts(opts, src);
  expectTrimmedMatch(linked, expected);
});

Deno.test("lower 5 bindings", async () => {
  const src = `
    struct Uniforms {
      foo: u32
    }

    struct MyBindings {
      @group(0) @binding(0) particles: ptr<storage, array<u32>, read_write>, 
      @group(0) @binding(1) uniforms: ptr<uniform, Uniforms>, 
      @group(0) @binding(2) tex: texture_2d<rgba8unorm>,
      @group(0) @binding(3) samp: sampler,
      @group(0) @binding(4) stTex: texture_storage_2d<rgba8unorm, read>,
    }

    @compute fn main(b: MyBindings) {
      b.particles[0] = b.uniforms.foo;
    }
  `;

  const expected = `
@group(0) @binding(0) var<storage, read_write> particles : array<u32>;
@group(0) @binding(1) var<uniform> uniforms : Uniforms;
@group(0) @binding(2) var tex : texture_2d<rgba8unorm>;
@group(0) @binding(3) var samp : sampler;
@group(0) @binding(4) var stTex : texture_storage_2d<rgba8unorm, read>;

    struct Uniforms { foo: u32 }
    @compute fn main() {
      particles[0] = uniforms.foo;
    }
`;

  const opts = { config: { plugins: [bindingStructsPlugin()] } };
  const linked = await linkTestOpts(opts, src);
  expectTrimmedMatch(linked, expected);
});
