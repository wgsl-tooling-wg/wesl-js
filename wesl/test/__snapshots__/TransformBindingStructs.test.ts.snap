export const snapshot = {};

snapshot[`transformBindingStruct 1`] = `
"@group(0) @binding(0) var<storage, read_write> particles : array<f32>;
"
`;

snapshot[`findRefsToBindingStructs 1`] = `
"memberRef b.particles
  ref b
  text '.'
  name particles"
`;

snapshot[`transformBindingReference 1`] = `"synthetic 'particles'"`;

snapshot[`lower binding structs 1`] = `
"module
  synthetic '@group(0) @binding(0) var<storage, read_write> particles : array<f32>;
'
  text '
    '
  text '
    '
  fn main(b: Bindings)
    decl %main
    param
    statement
      text '{
      let '
      typeDecl %x
        decl %x
      text ' = '
      memberRef b.particles
        synthetic 'particles'
      text ';
    }'
  text '
  '"
`;
